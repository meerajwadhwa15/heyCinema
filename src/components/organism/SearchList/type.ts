export type listType = {
  imdbID: string;
  Title: string;
  Year: string;
  Type: string;
  Poster: string;
};

export interface ComponentProps {}

export interface ComponentState {
  searchString: string;
  page: number;
}

export interface ReducerState {
  list: listType[];
  isFetching: boolean;
  errorMessage: string;
  noMoreRecords: boolean;
}

export interface DispatchProps {
  fetchSearchListAction: (searchString: string, page: number) => void;
  loadMoreSearchListAction: (searchString: string, page: number) => void;
}

export interface SearchListReducerType {
  SearchListReducer: ReducerState;
}

export interface ReducerAction {
  type: string;
  list: listType[];
  isFetching: boolean;
  message: string;
}

export type FetchSuccessActionType = {
  type: string;
  list: listType[];
};

export interface SetFetch {
  type: string;
  isFetching: boolean;
}

export type FetchErrorActionType = {
  type: string;
  message: string;
};

export type IntersectionObserverEntryType = {
  intersectionRatio: number;
};

export type IntersectionObserverEntriesType = IntersectionObserverEntryType[];
