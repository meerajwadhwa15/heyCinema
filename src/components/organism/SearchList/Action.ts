import { AnyAction } from 'redux';
import { AxiosResponse, AxiosError } from 'axios';
import { FetchSuccessActionType, SetFetch, FetchErrorActionType } from './type';
import { ThunkDispatch } from 'redux-thunk';
import Fetch, { ServerProps } from './../../../libs/fetch';
import { ERROR_MESSAGE } from './../../../config/message';
import { SEARCH_MOVIES } from './../../../config/api';

export const isFetching = (isFetching: boolean): SetFetch => {
  return { type: 'SET_FETCHING', isFetching };
};

export const FetchBeers = (searchString: string) => {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>): Promise<void> => {
    const showLoader: SetFetch = isFetching(true);
    dispatch(showLoader);
    const serverProps: ServerProps = {
      url: SEARCH_MOVIES,
      params: {
        s: searchString
      }
    };

    return Fetch(serverProps)
      .then((response: AxiosResponse) => {
        const fetchBeersSuccess: FetchSuccessActionType = {
          type: 'FETCH_MOVIES_SUCCESS',
          list: response.data.Search || []
        };
        dispatch(fetchBeersSuccess);
      })
      .catch((err: AxiosError) => {
        const fetchBeersSuccess: FetchErrorActionType = {
          type: 'FETCH_MOVIES_ERROR',
          message: ERROR_MESSAGE
        };
        dispatch(fetchBeersSuccess);
      })
      .finally(() => {
        const hideLoader: SetFetch = isFetching(false);
        dispatch(hideLoader);
      });
  };
};
