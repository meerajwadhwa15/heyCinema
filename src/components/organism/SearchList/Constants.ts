export enum Types {
  SET_FETCHING = 'SET_FETCHING',
  FETCH_LOAD_MORE_MOVIES_SUCCESS = 'FETCH_LOAD_MORE_MOVIES_SUCCESS',
  FETCH_LOAD_MORE_MOVIES_ERROR = 'FETCH_LOAD_MORE_MOVIES_ERROR',
  FETCH_MOVIES_SUCCESS = 'FETCH_MOVIES_SUCCESS',
  FETCH_MOVIES_ERROR = 'FETCH_MOVIES_ERROR'
}
