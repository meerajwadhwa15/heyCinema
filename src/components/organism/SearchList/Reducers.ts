import { Reducer } from 'redux';
import { ReducerState, ReducerAction } from './type';

export const initialState: ReducerState = {
  list: [],
  isFetching: false,
  errorMessage: ''
};

const SearchListReducer: Reducer<ReducerState, ReducerAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case 'SET_FETCHING':
      return { ...state, isFetching: action.isFetching };
    case 'FETCH_MOVIES_SUCCESS':
      return { ...state, errorMessage: '', list: action.list || [] };
    case 'FETCH_MOVIES_ERROR':
      return { ...state, errorMessage: action.message };
    default:
      return state;
  }
};

export default SearchListReducer;
