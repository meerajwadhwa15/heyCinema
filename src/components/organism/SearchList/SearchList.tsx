import React, { ReactNode, Component, Fragment, SyntheticEvent } from 'react';
import { connect } from 'react-redux';
import { ThunkDispatch } from 'redux-thunk';
import ListCard from '../../molecules/ListCard';
import SearchForm from './../../molecules/SearchForm';
import {
  listType,
  ComponentState,
  SearchListReducerType,
  ReducerState,
  DispatchProps,
  ComponentProps
} from './type';
import { FetchSearchList } from './Action';
import Loader from '../../molecules/Loader';
import ErrorMessage from '../../atoms/ErrorMessage';
import NoRecordFound from '../../atoms/NoRecordFound';

export const mapStateToProps = ({
  SearchListReducer: { list, isFetching, errorMessage }
}: SearchListReducerType): ReducerState => ({
  list,
  isFetching,
  errorMessage
});

export const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  fetchSearchListAction: (searchString: string) => {
    dispatch(FetchSearchList(searchString));
  }
});

export type Props = ReducerState & DispatchProps;

export class SearchList extends Component<Props, ComponentState> {
  static defaultProps = {
    list: []
  };

  handleSearch(value: string) {
    if (value) {
      const { fetchSearchListAction } = this.props;
      fetchSearchListAction(value);
    }
  }

  render(): ReactNode {
    const { list, isFetching, errorMessage } = this.props;

    return (
      <Fragment>
        <SearchForm searchResults={this.handleSearch.bind(this)} />
        <div className="cards">
          {isFetching ? (
            <Loader />
          ) : errorMessage ? (
            <ErrorMessage>{errorMessage}</ErrorMessage>
          ) : list.length ? (
            list.map((item: listType) => (
              <ListCard key={item.imdbID} item={item} />
            ))
          ) : (
            <NoRecordFound>No Record Found.</NoRecordFound>
          )}
        </div>
      </Fragment>
    );
  }
}

export default connect<
  ReducerState,
  DispatchProps,
  ComponentProps,
  SearchListReducerType
>(
  mapStateToProps,
  mapDispatchToProps
)(SearchList);
