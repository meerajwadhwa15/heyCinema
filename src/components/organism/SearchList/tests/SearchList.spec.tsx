import React from 'react';
import { shallow } from 'enzyme';
import { SearchList, mapStateToProps, Props } from '../SearchList';
import { create, ReactTestInstance } from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';

const sampleLoadingMockData: Props = {
  fetchSearchListAction: jest.fn(),
  list: [],
  isFetching: true
};

const sampleListMockData: Props = {
  fetchSearchListAction: jest.fn(),
  list: [
    {
      imdbID: '1',
      Title: 'Title 1',
      Year: '2019',
      Type: 'Type 1',
      Poster: 'Poster_1'
    },
    {
      imdbID: '2',
      Title: 'Title 2',
      Year: '2020',
      Type: 'Type 2',
      Poster: 'Poster_2'
    }
  ],
  isFetching: false
};

const ComponentWithLoading = () => (
  <BrowserRouter>
    <SearchList {...sampleListMockData} />
  </BrowserRouter>
);

const ComponentWithList = () => (
  <BrowserRouter>
    <SearchList {...sampleLoadingMockData} />
  </BrowserRouter>
);

describe('SearchList Enzyme testing', () => {
  it('renders with Loading without crashing', () => {
    shallow(<ComponentWithLoading />);
  });

  it('renders with SearchList without crashing', () => {
    shallow(<ComponentWithList />);
  });

  test('Matches the snapshot with search List', () => {
    const component = create(<ComponentWithList />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('Matches the snapshot without search list', () => {
    const component = create(<ComponentWithLoading />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('Should call fetchSearchListAction ', () => {
    const spy = jest.spyOn(sampleLoadingMockData, 'fetchSearchListAction');
    const component = create(<SearchList {...sampleLoadingMockData} />);
    const instance: ReactTestInstance = component.getInstance();
    instance.handleSearch({
      preventDefault: jest.fn
    });
    expect(spy).toBeCalled();
  });

  test('Should test mapStateToProps', () => {
    expect(
      mapStateToProps({
        SearchListReducer: {
          list: [],
          isFetching: false,
          errorMessage: ''
        }
      })
    ).toEqual({
      list: [],
      isFetching: false,
      errorMessage: ''
    });
  });
});
