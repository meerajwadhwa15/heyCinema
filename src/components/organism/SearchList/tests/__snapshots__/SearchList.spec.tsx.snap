// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchList Enzyme testing Matches the snapshot with search List 1`] = `
Array [
  <div
    className="search-form"
  >
    <form>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search"
        type="text"
        value=""
      />
    </form>
  </div>,
  <div
    className="cards"
  >
    <div
      className="loader"
    >
      Loading...
    </div>
  </div>,
]
`;

exports[`SearchList Enzyme testing Matches the snapshot without search list 1`] = `
Array [
  <div
    className="search-form"
  >
    <form>
      <input
        onBlur={[Function]}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="Search"
        type="text"
        value=""
      />
    </form>
  </div>,
  <div
    className="cards"
  >
    <article
      className="card"
    >
      <a
        href="/1"
        onClick={[Function]}
        title="Title 1"
      >
        <div>
          <img
            alt="Title 1"
            src="Poster_1"
            width="200"
          />
        </div>
        <div>
          <h3>
            Title 1
          </h3>
          <p>
            Type 1
          </p>
          <p>
            2019
          </p>
        </div>
      </a>
    </article>
    <article
      className="card"
    >
      <a
        href="/2"
        onClick={[Function]}
        title="Title 2"
      >
        <div>
          <img
            alt="Title 2"
            src="Poster_2"
            width="200"
          />
        </div>
        <div>
          <h3>
            Title 2
          </h3>
          <p>
            Type 2
          </p>
          <p>
            2020
          </p>
        </div>
      </a>
    </article>
  </div>,
]
`;
